userspace/libsinsp/parsers.cpp

```diff
@@ -4232,7 +4244,13 @@ void sinsp_parser::parse_rw_exit(sinsp_evt *evt)
 				// If the fd still doesn't contain tuple info (because the socket is a datagram one or because some event was lost),
 				// add it here.
 				//
-				if(!retrieve_enter_event(enter_evt, evt))
+
+				// sendmmsg has all data in the exit event itself.
+				if (etype == PPME_SOCKET_SENDMMSG_X)
+				{
+					enter_evt = evt;
```

Comment:
I'm curious, why is this a safe thing to do, and why not to retried the enter event? If an enter and exit event will contains the same data that will for sure break some assumptions down the line. An innocent example of what I'm thinking about is the latency calculation, where the timestamps are compared.

Reply:
Good question! First off, I had to change how this syscall works, usually Falco expects their syscalls to be like:
```
enter
exit
enter
exit
...
``` 
Because I had to loop over messages and send multiple events, this ended up looking like:
```
enter
enter
enter
...
exit
exit
exit
...
```
This broke assumptions made by Falco and resulted in weird behavior like assigning the same IP tuple to all events if the tuple was sent from the enter event. In order to work around this, I had to make sendmmsg submit empty events in the enter event and handle all arguments in the exit event (see [here](https://github.com/stackrox/falcosecurity-libs/blob/985c1c1cf3ba4e1fee92919f845d2238f66a2751/driver/event_table.c#L103-L104)). The recvmmsg is slightly different because it only issues the fd as the arguments for the enter event and this is correct, because that fd is actually used for all messages, so it goes:
```
enter
exit
exit
exit
...
```

Now, why is _this_ line in particular safe? Well, it's only used in two places before returning:
https://github.com/stackrox/falcosecurity-libs/blob/985c1c1cf3ba4e1fee92919f845d2238f66a2751/userspace/libsinsp/parsers.cpp#L4258
If you look at the `update_fd` definition, the second parameter is a const pointer and, looking at the implementation, copies of the required fields are added to the event.

https://github.com/stackrox/falcosecurity-libs/blob/985c1c1cf3ba4e1fee92919f845d2238f66a2751/userspace/libsinsp/parsers.cpp#L4289
Here, `evt->get_fd_info()->m_name` is a `std::string`, so the assignment will make a copy of the 0-th element returned by `enter_evt->get_param_as_str(...)`, so it's also safe.

Any other handling specific to sendmmsg is addressed in this parsing function and some of it in the `sinsp_parser::reset` method ([here](https://github.com/stackrox/falcosecurity-libs/blob/985c1c1cf3ba4e1fee92919f845d2238f66a2751/userspace/libsinsp/parsers.cpp#L509))
https://github.com/stackrox/falcosecurity-libs/blob/985c1c1cf3ba4e1fee92919f845d2238f66a2751/userspace/libsinsp/parsers.cpp#L796-L800

As for the latency calculation... Yeah, that is gonna be broken for this syscall, as stated at the beginning of this mile long comment, the enter/exit event assumption from Falco has to be broken for these syscalls to be tracked. If you have an idea on how we could unbreak it, that'd be great.